Update ddl.sql

trevorcummingsCCS opened the pull request:
The first cut database script.

trevorcummingsCCS committed:
Update ddl.sql

The first cut database script.

tommyb82 reviewed (commented):
Initial comment - I think we'll need to split this between Agreements and GM tables etc - they will be two separate databases 👍

tommyb82 commented on guided-match/ddl.sql:
```diff
@@ -0,0 +1,200 @@
+Description This file is a script to create the tables for the Find a Thing database. The data model for FaT will be deployed in 
+            two separate databases. One will be a postgreSQL database for our relational tables and the other will be a Neo4j
+            for the graph database.
+            
```
Test comment..

trevorcummingsCCS committed:
Update ddl.sql

Commented oput table which should be in the agreement service

trevorcummingsCCS committed:
Update ddl.sql

Moved create statements for commercial_agreements and lots from the FaT ddl script to the agreements ddl script.

trevorcummingsCCS committed:
Update ddl.sql

Fixed a  spelling mistake

trevorcummingsCCS commented:
Hi Adrian & Tom, I have made those further changes as requested. Trevor

tommyb82 commented on agreements/ddl.sql:
```diff
@@ -0,0 +1,41 @@
+            
+*/
+
+CREATE TABLE commercial_agreements (
```
Few missing fields / related tables perhaps according to latest [API defintion](https://raw.githubusercontent.com/Crown-Commercial-Service/ccs-scale-api-definitions/master/agreements/agreements-service.yaml):

- Agreement -> `benefits` (list of strings)
- Lot -> `description` (string)
- Lot -> `type` (enum `product|service|product and service`)
- Lot -> `sectors` (list of strings)
- Lot -> `routesToMarket`, `buyerNeeds` and `rules` (complex types, see API spec for detail)

Easiest way to visualise the API spec is to copy the entire thing into the online [Swagger Editor](https://editor.swagger.io/) and navigate through the Schemas section on the right hand side.

tommyb82 replied:
I hasten to add - we don't need all this yet.. 👍

trevorcummingsCCS replied:
Are they missing in the agreements ddl? The agreements DDL is just holding what is needed for FaT and not the agreements service. The data model for the Agreement service is here https://app.lucidchart.com/documents/edit/ffb89d3c-ee65-4ab2-b5cb-e42ca8213204/0_0

trevorcummingsCCS replied:
I think we need to agree on one communication channel as Google Hangouts, Slack, email and GitHub is sending my head into a spin.

tommyb82 replied:
Github primarily for review comments and discussion on code (with further expansion on Slack as required) 👍

adrianmilne replied:
> I think we need to agree on one communication channel as Google Hangouts, Slack, email and GitHub is sending my head into a spin.

I can set up a WhatsApp group to discuss further?

tommyb82 replied:
To clarify requirements for FaT Phase 1, we need the following additional columns at Lot level I believe (from the API):

- `type` (enum `bat`, `cat`)
- `routeToMarket` (enum `fc`, `da`)
- `scale` (boolean)

tommyb82 commented on guided-match/ddl.sql:
```diff
@@ -0,0 +1,204 @@
+CREATE TABLE search_domains (
+  domain_modifier_id               INTEGER PRIMARY KEY,
+  search_id                        INTEGER NOT NULL,
+  journey_id                       INTEGER NOT NULL,
```
`journey_id` needs to accommodate a UUID value for linking to the graph.

trevorcummingsCCS replied:
Fixed all tables which reference journey_id to be of type UUID

tommyb82 commented on guided-match/ddl.sql:
```diff
@@ -0,0 +1,204 @@
+
+CREATE TABLE search_domains (
+  domain_modifier_id               INTEGER PRIMARY KEY,
+  search_id                        INTEGER NOT NULL,
```
This is a foreign key from `search_terms` presumably - formalise this using `references`

tommyb82 commented on guided-match/ddl.sql:
```diff
@@ -0,0 +1,204 @@
+
+/* This table could exist in some form in both databases */
+CREATE TABLE journeys (
+  journey_id                        INTEGER PRIMARY KEY,
```
Probably need to add a `UUID` type column to hold the reference to the Journey in the graph

trevorcummingsCCS replied:
Have changed from INTEGER to UUID

tommyb82 commented on guided-match/ddl.sql:
```diff
@@ -0,0 +1,204 @@
+
+CREATE TABLE journey_instance_questions (
+  journey_instance_question_id      INTEGER PRIMARY KEY,
+  journey_question_id               INTEGER NOT NULL,
```
Question - is `journey_question_id` here meant to be `journey_instance_id` (and a foreign key back to `journey_instances`)?  Probably also need:

- `UUID` type column to accommodate a reference to the question
- `hint` (text)
- This model does not currently appear to accommodate multi-select questions (but the service, graph and [API](https://raw.githubusercontent.com/Crown-Commercial-Service/ccs-scale-api-definitions/master/guided-match/guided-match-service.yaml) do - see `QuestionHistory` and `AnswerHistory` schema types)
- `answer_id` needs to be type `UUID` to reference the corresponding answer node in the graph

trevorcummingsCCS replied:
Yes needs a fk back to journey instance. Will add.

trevorcummingsCCS replied:
Not sure what hint relates to. Will check with you. I have added fk to journey_iinstances and made it type UUID. I have chnaged type of answer_id to UUID and changed the type of journey_question_id to UUID.

Regarding multi select questions I will have to think about this. I think it needs an extra table.

trevorcummingsCCS replied:
Talking to Dave Hughes we don't think that this table needs a hint column as this table is a history table. I do agree about multiple answers and need to look into this.

tommyb82 replied:
> Talking to Dave Hughes we don't think that this table needs a hint column as this table is a history table. I do agree about multiple answers and need to look into this.

Agreed - must have been coffee deprived when I put that comment in!

tommyb82 commented on guided-match/ddl.sql:
```diff
@@ -0,0 +1,204 @@
+*/
+
+/* This table creation script belongs in the agreement ddl.sql
+CREATE TABLE commercial_agreements (
```
Remove this `commercial_agreements` commented block

trevorcummingsCCS replied:
Done

tommyb82 reviewed (changes requested):
Few questions / points to clarify - thanks Trev 👍

tommyb82 commented on agreements/ddl.sql:
```diff
@@ -0,0 +1,41 @@
+  lot_id                            INTEGER PRIMARY KEY,
+  lot_number                        VARCHAR(20) NOT NULL,
+  lot_name                          VARCHAR(200) NOT NULL,
+  commercial_agreement_id           INTEGER NOT NULL,
```
Foreign key?

trevorcummingsCCS replied:
Done

adrianmilne commented on agreements/ddl.sql:
```diff
@@ -0,0 +1,41 @@
+);
+
+CREATE INDEX COAG_IDX1 on COMMERCIAL_AGREEMENTS  (commercial_agreement_number);
+CREATE INDEX COAG_IDX2 on COMMERCIAL_AGREEMENTSS (commercial_agreement_name);
```
Typo - 'SS'

trevorcummingsCCS replied:
Done

adrianmilne commented on guided-match/ddl.sql:
```diff
@@ -0,0 +1,204 @@
+  search_term                      VARCHAR(50) NOT NULL
+);
+
+CREATE INDEX SETE_IDX1 ON SEARCH_TERMS (serach_term);
```
Typo - 'serach_term'

trevorcummingsCCS replied:
Fixed this

adrianmilne commented on guided-match/ddl.sql:
```diff
@@ -0,0 +1,204 @@
+  answer_id                         INTEGER NOT NULL,
+  answer_text                       VARCHAR(2000) 
+);
+CREATE INDEX JOIQ_IDX1 on JOURNEY_INSTANCE_QUESTIONS (journey_questiuon_id);
```
typo - 'journey_questiuon_id'

trevorcummingsCCS replied:
fixed this

trevorcummingsCCS commented:
Hi Adrian.

Apologies for typos.

Did you fix these in GitHub? if not I'll do them this morning. When I have
time I will set up an instance of Postgres, then I can test these scripts.

Thanks

trev

On Thu, 28 May 2020 at 02:14, Adrian Milne <notifications@github.com> wrote:

> *@adrianmilne* commented on this pull request.
> ------------------------------
>
> In guided-match/ddl.sql
> <https://github.com/Crown-Commercial-Service/ccs-scale-db-scripts/pull/2#discussion_r431528951>
> :
>
> > +  default_ind                      CHAR(1), -- Y or N default value will be N as only one can be Y
> +  valid_value                      VARCHAR(200),
> +  url                              VARCHAR(200), -- notsure required anymore
> +  regex_validation                 VARCHAR(2000)  -- notsure required anymore
> +);
> +*/
> +
> +CREATE TABLE journey_instance_questions (
> +  journey_instance_question_id      INTEGER PRIMARY KEY,
> +  journey_question_id               INTEGER NOT NULL,
> +  question_order                    INTEGER NOT NULL,
> +  question_text                     VARCHAR(2000) NOT NULL,
> +  answer_id                         INTEGER NOT NULL,
> +  answer_text                       VARCHAR(2000)
> +);
> +CREATE INDEX JOIQ_IDX1 on JOURNEY_INSTANCE_QUESTIONS (journey_questiuon_id);
>
> typo - 'journey_questiuon_id'
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Crown-Commercial-Service/ccs-scale-db-scripts/pull/2#pullrequestreview-419713516>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AOJORTDXDEHBL7KTQEUJINDRTW3JFANCNFSM4NGYCVWQ>
> .
>


-- 



[image: Description: cid:image003.png@01CF91F1.B7B56CE0]
Trevor Cummings

Data Architect

Crown Commercial Service

151 Buckingham Palace Road, London, SW1W 9SS


Tel: 07771 544355

www.crowncommercial.gov.uk  | follow us on Twitter
<https://twitter.com/gov_procurement> | connect with us on LinkedIn
<http://www.linkedin.com/company/government-procurement-service?trk=cp_followed_logo_government-procurement-service>


*Crown Commercial Service - delivering value for the nation through
outstanding commercial capability and quality customer service*

adrianmilne commented:
Hi @trevorcummingsCCS - no I haven't changed them - the usual process is for the person who requested the pull request to make all changes - so they have control/oversight of all changes (and also the option to rebuke them).

You can click the 'Resolve Conversation' button under each comment as the issue is addressed - which can make it easier to keep track of things I find if there are a number of scattered comments (then you just need to look at the unresolved comments until all are resolved).

Once we have ironed out some issues with our DB creation/access you may be able to use one of those. I am using DB Visualiser - which will highlight errors in a script without the need to execute it (but you do need a valid connection enabled).

trevorcummingsCCS committed:
Update ddl.sql

Fixed typos

trevorcummingsCCS committed:
I've fixed typos and added sectors

trevorcummingsCCS committed:
Removed commented blocks for Commercial Agreements & Lots

trevorcummingsCCS committed:
Changed journey_id from type integer to type uuid

trevorcummingsCCS committed:
Added FKS to lots and lot_sectors

trevorcummingsCCS committed:
Have added FK back to Journey_instance from journey_instance_questions

Have chnaged the types where they were INTEGER to UUID where they reference back to objects in the Neo4J database.

adrianmilne commented on guided-match/ddl.sql:
```diff
@@ -0,0 +1,176 @@
+
+CREATE TABLE journey_instance_questions (
+  journey_instance_question_id      INTEGER PRIMARY KEY,
+  journey_instance_id               UUID NOT NULL
```
Missing comma at end of line

trevorcummingsCCS replied:
Fixed .

adrianmilne commented on guided-match/ddl.sql:
```diff
@@ -0,0 +1,176 @@
+  answer_text                       VARCHAR(2000) 
+);
+CREATE INDEX JOIQ_IDX1 on JOURNEY_INSTANCE_QUESTIONS (journey_question_id);
+CREATE INDEX JOIQ_IDX2 on JOURNEY_INSTANCE_QUESTIONS (answer_id);
```
Complains that relation 'journey_instance_questions' does not exist (on these 2 lines)

adrianmilne replied:
Sorry - forget this - it is fine - must have been a side effect of the problem above - goes away when that is fixed

trevorcummingsCCS committed:
added missing comma after journey_instance_id  UUID NOT NULL

tommyb82 commented on agreements/ddl.sql:
```diff
@@ -0,0 +1,71 @@
+    
+ALTER TABLE lot_sectors 
+ADD CONSTRAINT lot_sectors_sectors_fk FOREIGN KEY (sector_code) 
+    REFERENCES lots (sector_code);        
```
Should be table `sectors`?

trevorcummingsCCS replied:
Fixed that. I also have added a drop table script but I am having GitHub issues as I tried to create a new Pull Request to ask you guys to review it.

tommyb82 commented on agreements/ddl.sql:
```diff
@@ -0,0 +1,71 @@
+CREATE INDEX LOTS_IDX3 on LOTS(commercial_agreement_id);
+
+CREATE TABLE sectors (
+  sector_code                       VARCHAR(2) NOT NULL,
```
`PRIMARY KEY,`

trevorcummingsCCS replied:
Have just fixed that. I have been away from writing scripts for too long

trevorcummingsCCS committed:
File to drop tables for teh Agreements service

tommyb82 reviewed (commented):
@trevorcummingsCCS - Just noticed a couple of little required fixes on the agreements DDL, thought I'd mention it now whilst I'm here (I know you're still working on the GM DDL following conversation yesterday 👍

trevorcummingsCCS committed:
Changed lots to sectors

trevorcummingsCCS committed:
Amended tables for missing Primary key definition.

trevorcummingsCCS committed:
Created drop table script for Find a Thing service.

trevorcummingsCCS committed:
Added Fks

trevorcummingsCCS committed:
Added table journey_instance_answers 

This meant removing two columns from journey_instance_questions.

tommyb82 commented on guided-match/ddl.sql:
```diff
@@ -0,0 +1,207 @@
+  answer_sequence                   SMALLINT,
+  answer_text                       VARCHAR(2000),
+  answer_date                       DATE,
+  answer_number                     NUMBER(18,4)
```
`NUMERIC`

trevorcummingsCCS replied:
Changed to numeric. Must have been on Oracle auto-pilot

tommyb82 commented on guided-match/ddl.sql:
```diff
@@ -0,0 +1,207 @@
+ADD CONSTRAINT journey_instance_questions_journey_instances_fk FOREIGN KEY (journey_instance_id) 
+    REFERENCES journey_instances (journey_instance_id); 
+    
+ALTER TABLE journey_instance_questions 
```
`journey_question_id` is a UUID from the graph - no FK required here

trevorcummingsCCS replied:
Removed fk refrence that wasn't needed.

tommyb82 replied:
I think this should be:
```
ALTER TABLE journey_instance_questions 
ADD CONSTRAINT journey_instance_questions_journey_instance_fk FOREIGN KEY (journey_instance_id) 
    REFERENCES journey_instances (journey_instance_id);
```
?

tommyb82 commented on guided-match/drop_tables.sql:
```diff
@@ -0,0 +1,20 @@
+Description This file is a script to drop the tables for the Find a Thing database. 
+*/
+
+DROP TABLE journey_instance_questions ;
```
Requires `DROP TABLE journey_instance_answers;` first

trevorcummingsCCS replied:
Added drop table statement.

tommyb82 commented on guided-match/ddl.sql:
```diff
@@ -0,0 +1,207 @@
+    
+ALTER TABLE search_domains
+ADD CONSTRAINT search_domains_journeys_fk FOREIGN KEY (journey_id) 
+    REFERENCES search_terms(journey_id);   
```
[Code: 0, SQL State: 42703]  ERROR: column "journey_id" referenced in foreign key constraint does not exist

trevorcummingsCCS replied:
Changed refrence to journeys from search_terms to journeys.

tommyb82 reviewed (changes requested):
Few more fixes required 👍

trevorcummingsCCS committed:
Fixed typos again

In journey_instance_answers chnaged number to numeric, old oracle ways die hard.
Removed constraint to journey_instances which is in the Neo4J database.
Amended FK from search_domains to journeys

trevorcummingsCCS committed:
Added drop journey_instance_answers.

tommyb82 commented on guided-match/drop_tables.sql:
```diff
@@ -0,0 +1,22 @@
+
+DROP TABLE journeys ;
+
+DROP TABLE search_domains ;
```
Need to delete `search_domains` before `journeys`

tommyb82 reviewed (changes requested):
Final two! Then it all works 👍

trevorcummingsCCS committed:
Amended order of drops

trevorcummingsCCS committed:
Amended the fk

tommyb82 reviewed (approved):
All working now - great, thanks @trevorcummingsCCS 👍
